src/native/simplekdf.rs

```diff
@@ -0,0 +1,280 @@
+    /// The additional key in CKM_CONCATENATE_BASE_AND_KEY
+    key: Option<KeyInfo>,
+    /// The additional data to concatenate
+    data: Option<Vec<u8>>,
+}
+
+unsafe impl Send for SimpleKDFOperation {}
+unsafe impl Sync for SimpleKDFOperation {}
+
+impl SimpleKDFOperation {
+    /// Creates a new Simple KDF operation based on the provided mechanism
+    ///
+    /// Dispatches to specific constructors based on the mechanism type.
+    pub fn new(mech: &CK_MECHANISM) -> Result<SimpleKDFOperation> {
+        match mech.mechanism {
+            CKM_CONCATENATE_BASE_AND_KEY => {
+                let object_handle = cast_params!(mech, CK_OBJECT_HANDLE);
+                Ok(SimpleKDFOperation {
+                    finalized: false,
+                    mech: mech.mechanism,
+                    key_handle: Some([object_handle]),
+                    key: None,
+                    data: None,
+                })
+            }
+            CKM_CONCATENATE_BASE_AND_DATA
+            | CKM_CONCATENATE_DATA_AND_BASE
+            | CKM_XOR_BASE_AND_DATA => {
+                let params = cast_params!(mech, CK_KEY_DERIVATION_STRING_DATA);
+
+                let data = bytes_to_vec!(params.pData, params.ulLen);
+                Ok(SimpleKDFOperation {
+                    finalized: false,
+                    mech: mech.mechanism,
+                    key_handle: None,
+                    key: None,
+                    data: Some(data),
+                })
+            }
+            _ => return Err(CKR_MECHANISM_INVALID)?,
+        }
+    }
+}
+
+impl MechOperation for SimpleKDFOperation {
+    fn mechanism(&self) -> Result<CK_MECHANISM_TYPE> {
+        Ok(self.mech)
+    }
+
+    fn finalized(&self) -> bool {
+        self.finalized
+    }
+    #[cfg(feature = "fips")]
+    fn fips_approved(&self) -> Option<bool> {
+        Some(true)
+    }
+    fn requires_objects(&self) -> Result<&[CK_OBJECT_HANDLE]> {
+        if self.mech == CKM_CONCATENATE_BASE_AND_KEY {
+            match &self.key_handle {
+                Some(h) => Ok(h),
+                None => Err(CKR_OK)?,
+            }
+        } else {
+            Err(CKR_OK)?
+        }
+    }
+    fn receives_objects(&mut self, objs: &[&Object]) -> Result<()> {
+        if objs.len() != 1 {
+            return Err(CKR_GENERAL_ERROR)?;
+        }
+        self.key = Some(KeyInfo::new_from_object(objs[0])?);
+        Ok(())
+    }
+}
+
+impl Derive for SimpleKDFOperation {
+    /// Performs the appropriate Simple KDF derivation based on the mechanism used
+    fn derive(
+        &mut self,
+        key: &Object,
+        template: &[CK_ATTRIBUTE],
+        _mechanisms: &Mechanisms,
+        objfactories: &ObjectFactories,
+    ) -> Result<Vec<Object>> {
+        if self.finalized {
+            return Err(CKR_OPERATION_NOT_INITIALIZED)?;
+        }
+        self.finalized = true;
+
+        key.check_key_ops(CKO_SECRET_KEY, CKK_GENERIC_SECRET, CKA_DERIVE)?;
```

Comment:
In the fips indicator you leave the door open to derive any key type, but here you are forcing CKK_GENERIC_SECRET as the only valid one.

Does the spec allow only CKK_GENERIC_SECRET keys in input?
Perhaps you should use CK_UNAVAILABLE_INFORMATION here to allow any key type?

Reply:
I think the specs does not restrict the key types so lets use UNAVAILABLE_INFORMATION to match any key type. Generally I think AES AND GENERIC_SECRET ones though.